src/Microsoft.OData.Core/ODataPayloadValueConverter.cs

```diff
@@ -152,6 +152,10 @@ private static object ConvertStringValue(string stringValue, Type targetType)
 
             if (targetType == typeof(Guid))
             {
+                if (stringValue == "null")
+                {
+                    return Guid.Empty;

```

Comment:
Out of curiosity, is Guid.Empty equivalent to it being null?

If you had class like 
```c#
class SomeEntity {
   int Id { get;set; }
   guid? SomeGuid {get;set}
}
```
if you set an object of this type like this
```
var entity = new SomeEntity();
entity.SomeGuid = null;
```
If you serialize and deserialize this, what value do we expect to have for entity.SomeGuid? null or Guid.Empty?

I assume if a field type is explicitly marked as nullable, then it would make sense to support null.

For nullable `int?` type, I assume 0 != null,

Reply:
This is a very good discussion you've started.  Also had some questions around this.

Reply:
@habbes  A Guid is a value type so it can't be null. You can  only check if its empty

Reply:
But in this case it's a nullable Guid, i.e. the type is `guid?` not jus a plan `guid`. A nullable value type can have `null` as the value: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types Maybe just to confirm, you could verify how we handle nullable ints?

Reply:
@xuzhg @mikepizzo.

Reply:
typically, it should return "return (Guid?)null";

But, my concern is, targetType is "Guid", not "Guid?"?

If that's the case, any other type has the similar problem? for example, TimeSpan?, Date?, etc.

So, why should we change the code here?  Shall we do the "null" check at other place?

---
In reply to: [603236521](https://github.com/OData/odata.net/pull/2035#discussion_r603236521) [](ancestors = 603236521)

Reply:
@xuzhg this means running this query http://host/service/TestEntity?$filter=NullableProperty in (null) should not return entities with empty Guids but with null guids? right?

Reply:
Curious, how does one filter for entities with empty Guids?
`?$filter=SomeProperty eq '00000000-0000-0000-0000-000000000000'` or is there some other way?

Reply:
@xuzhg  filtering using in operator with null  on nullable datetime, date etc works okay.